cpp/src/round/round.cu

```diff
@@ -0,0 +1,125 @@
+ */
+
+#include <cudf/column/column_factories.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/detail/round.hpp>
+#include <cudf/round.hpp>
+#include <cudf/scalar/scalar.hpp>
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+namespace cudf {
+
+namespace detail {
+
+struct round_fn {
+  template <typename T, typename... Args>
+  std::enable_if_t<not cudf::is_numeric<T>(), std::unique_ptr<column>> operator()(Args&&... args)
+  {
+    CUDF_FAIL("Type not currenlty support for cudf::round");
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_floating_point<T>(), std::unique_ptr<column>> operator()(
+    column_view const& input,
+    int32_t decimal_places,
+    cudf::round_option round,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    auto result   = cudf::make_fixed_width_column(input.type(), input.size());
+    auto out_view = result->mutable_view();
+    auto const n  = static_cast<int32_t>(std::pow(10, std::abs(decimal_places)));
+
+    if (decimal_places == 0)
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [] __device__(T e) -> T { return std::roundf(e); });
+    else if (decimal_places > 0)
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [n] __device__(T e) -> T { return std::roundf(e * n) / n; });
+    else  // decimal_places < 0
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [n] __device__(T e) -> T { return std::roundf(e / n) * n; });
+
+    return result;
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_integral<T>::value, std::unique_ptr<column>> operator()(
+    column_view const& input,
+    int32_t decimal_places,
+    cudf::round_option round,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream)
+  {
+    auto result = cudf::make_fixed_width_column(input.type(), input.size());
+
+    // only need to handle the case where decimal_places is < zero
+    // integers by definition have no fractional part, so result of "rounding" is a no-op
+    if (decimal_places < 0) {
+      auto out_view = result->mutable_view();
+      auto const n  = static_cast<int32_t>(std::pow(10, -decimal_places));
+
+      thrust::transform(rmm::exec_policy(stream)->on(stream),
+                        input.begin<T>(),
+                        input.end<T>(),
+                        out_view.begin<T>(),
+                        [n] __device__(T e) -> T { return std::roundf(1.0 * e / n) * n; });
```

Comment:
Is this going to do the right thing if we have an INT64?  If I have any int value that does not fit in the 23 bit fraction of a float combined with a decimal_places close to 0 (like -1) we would end up with a wrong answer because we lost information in the implicit cast to a float.

Reply:
I agree. Isn't it sufficient to just do `(e / n) * n`?

Reply:
This will require some custom logic for whether to round up or down, and also requires care near `std::numeric_limits<T>::max()` since the same arithmetic you use elsewhere will fail...